File: src/main/java/com/example/finsentinel/service/market/MarketCalendarService.java
package com.example.finsentinel.service.market;

import com.example.finsentinel.service.openbb.OpenbbPublicDataService;
import tools.jackson.databind.JsonNode;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.LinkedHashMap;
import java.util.Map;

@Service
@RequiredArgsConstructor
@Slf4j
public class MarketCalendarService {

    private final OpenbbPublicDataService openbbService;

    public String getUpcomingEarnings(String ticker) {
        return queryAndFormat("equity/calendar/earnings",
                Map.of("symbol", ticker.toUpperCase().trim()),
                "Upcoming Earnings for " + ticker.toUpperCase().trim());
    }

    public String getDividendHistory(String ticker) {
        return queryAndFormat("equity/calendar/dividend",
                Map.of("symbol", ticker.toUpperCase().trim()),
                "Dividend Calendar for " + ticker.toUpperCase().trim());
    }

    public String getSplitHistory(String ticker) {
        return queryAndFormat("equity/fundamental/historical_splits",
                Map.of("symbol", ticker.toUpperCase().trim()),
                "Stock Split History for " + ticker.toUpperCase().trim());
    }

    public String getIPOCalendar() {
        return queryAndFormat("equity/calendar/ipo", Map.of(), "Upcoming IPO Calendar");
    }

    private String queryAndFormat(String path, Map<String, String> params, String title) {
        try {
            JsonNode result = openbbService.queryPublicData(path, null, new LinkedHashMap<>(params));
            if (result == null || result.isEmpty()) {
                return title + "\nNo data available.";
            }
            StringBuilder sb = new StringBuilder();
            sb.append("=== ").append(title).append(" ===\n");
            if (result.isArray()) {
                int count = 0;
                for (JsonNode item : result) {
                    if (count >= 20) {
                        sb.append("... and ").append(result.size() - 20).append(" more entries\n");
                        break;
                    }
                    sb.append(item.toString()).append("\n");
                    count++;
                }
                sb.append("Total: ").append(result.size()).append(" entries\n");
            } else {
                sb.append(result.toPrettyString()).append("\n");
            }
            return sb.toString();
        } catch (Exception e) {
            log.warn("Calendar query failed for {}: {}", path, e.getMessage());
            return title + "\nData unavailable: " + e.getMessage();
        }
    }
}

File: apps/api/src/market/market-calendar.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { OpenbbPublicDataService } from '../openbb/openbb-public.service';

/**
 * Market calendar data (earnings, dividends, splits) via OpenBB Platform.
 *
 * All methods delegate to OpenbbPublicDataService and return raw JSON.
 */
@Injectable()
export class MarketCalendarService {
  private readonly logger = new Logger(MarketCalendarService.name);

  constructor(private readonly openbb: OpenbbPublicDataService) {}

  /**
   * Upcoming earnings calendar from OpenBB.
   *
   * @param startDate - optional start date (YYYY-MM-DD)
   * @param endDate   - optional end date (YYYY-MM-DD)
   */
  async getEarningsCalendar(
    startDate?: string,
    endDate?: string,
  ): Promise<unknown> {
    this.logger.debug(
      `Fetching earnings calendar: ${startDate ?? 'no start'} to ${endDate ?? 'no end'}`,
    );

    const params: Record<string, string> = {};
    if (startDate) params['start_date'] = startDate;
    if (endDate) params['end_date'] = endDate;

    return this.openbb.queryPublicData(
      'equity/calendar/earnings',
      undefined,
      params,
    );
  }

  /**
   * Upcoming dividend calendar from OpenBB.
   *
   * @param startDate - optional start date (YYYY-MM-DD)
   * @param endDate   - optional end date (YYYY-MM-DD)
   */
  async getDividendCalendar(
    startDate?: string,
    endDate?: string,
  ): Promise<unknown> {
    this.logger.debug(
      `Fetching dividend calendar: ${startDate ?? 'no start'} to ${endDate ?? 'no end'}`,
    );

    const params: Record<string, string> = {};
    if (startDate) params['start_date'] = startDate;
    if (endDate) params['end_date'] = endDate;

    return this.openbb.queryPublicData(
      'equity/calendar/dividend',
      undefined,
      params,
    );
  }

  /**
   * Upcoming stock splits calendar from OpenBB.
   *
   * @param startDate - optional start date (YYYY-MM-DD)
   * @param endDate   - optional end date (YYYY-MM-DD)
   */
  async getSplitsCalendar(
    startDate?: string,
    endDate?: string,
  ): Promise<unknown> {
    this.logger.debug(
      `Fetching splits calendar: ${startDate ?? 'no start'} to ${endDate ?? 'no end'}`,
    );

    const params: Record<string, string> = {};
    if (startDate) params['start_date'] = startDate;
    if (endDate) params['end_date'] = endDate;

    return this.openbb.queryPublicData(
      'equity/calendar/splits',
      undefined,
      params,
    );
  }

  async getUpcomingEarnings(ticker: string): Promise<unknown> {
    return this.openbb.queryPublicData(
      'equity/calendar/earnings',
      undefined,
      { symbol: ticker.toUpperCase() },
    );
  }

  async getDividendHistory(ticker: string): Promise<unknown> {
    return this.openbb.queryPublicData(
      'equity/calendar/dividend',
      undefined,
      { symbol: ticker.toUpperCase() },
    );
  }

  async getSplitHistory(ticker: string): Promise<unknown> {
    return this.openbb.queryPublicData(
      'equity/calendar/splits',
      undefined,
      { symbol: ticker.toUpperCase() },
    );
  }

  async getIPOCalendar(): Promise<string> {
    return 'IPO calendar is not available from the current OpenBB-backed market calendar service.';
  }
}
